app/components/new-directory-user.js

```diff
@@ -1,127 +1,208 @@
+
+    //prefetch programYears and programs so that ember data will coalesce these requests.
+    const programYears = await all(cohorts.getEach('programYear'));
+    await all(programYears.getEach('program'));
+
+    const objects = await all(
+      cohorts.toArray().map(async (cohort) => {
+        const obj = {
+          id: cohort.get('id'),
+        };
+        const programYear = await cohort.programYear;
+        const program = await programYear.program;
+        obj.title = program.title + ' ' + cohort.title;
+        obj.startYear = programYear.startYear;
+        obj.duration = program.duration;
+
+        return obj;
+      })
+    );
+
+    const lastYear = parseInt(moment().subtract(1, 'year').format('YYYY'), 10);
+    return objects.filter((obj) => {
+      const finalYear = parseInt(obj.startYear, 10) + parseInt(obj.duration, 10);
+      return finalYear > lastYear;
+    });
+  }
+
+  @restartableTask
+  *load() {
+    const authType = yield this.iliosConfig.authenticationType;
+    this.allowCustomUserName = 'form' === authType;
+    const user = yield this.currentUser.model;
+    this.primarySchool = yield user.school;
+    this.schools = yield this.loadSchools();
+    this.cohorts = yield this.loadCohorts(this.primarySchool);
+    this.currentSchoolCohorts = yield this.bestSelectedSchool?.cohorts;
+    if (isPresent(this.args.searchTerms)) {
+      yield this.findUsersInDirectory.perform(this.args.searchTerms);
     }
-  },
-
-  actions: {
-    pickUser(user) {
-      this.set('selectedUser', true);
-      this.set('firstName', user.firstName);
-      this.set('lastName', user.lastName);
-      this.set('email', user.email);
-      this.set('campusId', user.campusId);
-      this.set('phone', user.telephoneNumber);
-      this.set('username', user.username);
-    },
-
-    unPickUser() {
-      this.set('selectedUser', false);
-      this.set('firstName', null);
-      this.set('lastName', null);
-      this.set('email', null);
-      this.set('campusId', null);
-      this.set('phone', null);
-      this.set('username', null);
-    },
-  },
-
-  keyUp(event) {
+  }
+
+  @restartableTask
+  *reload() {
+    this.currentSchoolCohorts = yield this.bestSelectedSchool?.cohorts;
+    this.cohorts = yield this.loadCohorts(this.bestSelectedSchool);
+  }
+
+  @action
+  pickUser(user) {
+    this.selectedUser = true;
+    this.firstName = user.firstName;
+    this.lastName = user.lastName;
+    this.email = user.email;
+    this.campusId = user.campusId;
+    this.phone = user.telephoneNumber;
+    this.username = user.username;
+  }
+
+  @action
+  unPickUser() {
+    this.selectedUser = false;
+    this.firstName = null;
+    this.lastName = null;
+    this.email = null;
+    this.campusId = null;
+    this.phone = null;
+    this.username = null;
+  }
+
+  @action
+  setSchool(id) {
+    this.schoolId = id;
+  }
+
+  @action
+  setPrimaryCohort(id) {
+    this.primaryCohortId = id;
+  }
+
+  @action
+  async keyboard(event) {
```

Comment:
I'm not understanding why this method is async (and why the calls to `perform()` need to be awaited here and not just proceed after the keypress.

Reply:
:heavy_check_mark: